ts/src/connections/sshHelpers.ts

```diff
@@ -285,8 +306,74 @@ export class RelayConnectionError extends Error {
         message: string,
         public readonly errorContext: {
             errorType: RelayErrorType;
+            statusCode?: number;
         },
     ) {
         super(message);
     }
 }
+
+/**
+ * Web socket client error context.
+ */
+interface WebSocketClientErrorContext {
+    readonly regex?: RegExp;
+    readonly statusCode?: number;
+    readonly error: string;
+    readonly errorType: RelayErrorType;
+}
+
+/**
+ * Web socket client error contexts.
+ */
+
+// TODO: Return ProblemDetails from TunnelRelay service. The 404 error messages
+// below match Azure Relay but not TunnelRelay.
+// Status 404 may be used for at least two distinct error conditions.
+// So we have to match on the error message text. This could break when
+// the relay server behavior changes or when updating the client websocket library.
+// But then in the worst case the original error message will be reported.
+
+const webSocketClientContexts: WebSocketClientErrorContext[] = [
+    {
+        regex: /status: 401/,
+        statusCode: 401,
+        error: 'relayClientUnauthorized',
+        errorType: RelayErrorType.Unauthorized,
+    },
+    {
+        regex: /status: 403/,
+        statusCode: 403,
+        error: 'relayClientForbidden',
+        errorType: RelayErrorType.Unauthorized,
+    },
+    {
+        statusCode: 404,
+        error: 'tunnelPortNotFound',
+        errorType: RelayErrorType.TunnelPortNotFound,
+    },
+    {
+        regex: /status: 404 Endpoint does not exist/,
+        statusCode: 404,
+        error: 'relayEndpointNotFound',
+        errorType: RelayErrorType.EndpointNotFound,
+    },
+    {
+        regex: /status: 404 There are no listeners connected/,
+        statusCode: 404,
+        error: 'relayListenerOffline',
+        errorType: RelayErrorType.ListenerOffline,
+    },
```

Comment:
These error message patterns are leftover from old Azure Relay error-handling. I don't think they are valid any more for tunnel relay connections.

Reply:
And I see that's mentioned in the comments above. But it can be deleted now.